File: java-leetcode/src/main/java/dev/codefactory/leetcode/core/Solution0006ZigZagConversion.java
package dev.codefactory.leetcode.core;

public class Solution0006ZigZagConversion {

    public String convert(String s, int numRows) {
        if (s.length() <= 1) return s;

        /*

        PAYPALISHIRING

        PAHNAPLSIIGYIR

        [0] -> [0]
        [1] -> [4]
        [2] -> [11]
        [3] -> [5]
        [4] -> [6]

         */

        StringBuilder[] sb = new StringBuilder[numRows];
        for (int i = 0; i < numRows; i++) sb[i] = new StringBuilder();
        int i = 0, n = s.length();
        while (i < n) {
            for (int j = 0; j < numRows && i < n; j++)
                sb[j].append(s.charAt(i++));
            for (int j = numRows - 2; j > 0 && i < n; j--)
                sb[j].append(s.charAt(i++));
        }
        for (int j = 1; j < numRows; j++)
            sb[0].append(sb[j]);
        return sb[0].toString();
    }
}

File: kotlin-leetcode/src/main/kotlin/dev/codefactory/leetcode/core/Solution0006ZigZagConversion.kt
package dev.codefactory.leetcode.core

class Solution0006ZigZagConversion {

    fun convert(s: String, numRows: Int): String {
        if (s.length <= 1) return s

        val sb = arrayOfNulls<StringBuilder>(numRows)
        for (i in 0 until numRows) sb[i] = StringBuilder()
        var i = 0
        val n = s.length
        while (i < n) {
            run {
                var j = 0
                while (j < numRows && i < n) {
                    sb[j]!!.append(s[i++])
                    j++
                }
            }
            var j = numRows - 2
            while (j > 0 && i < n) {
                sb[j]!!.append(s[i++])
                j--
            }
        }
        for (j in 1 until numRows) sb[0]!!.append(sb[j])
        return sb[0].toString()
    }
}